esp-hal/src/dma/mod.rs

```diff
@@ -48,6 +48,137 @@
 
 use core::{fmt::Debug, marker::PhantomData, ptr::addr_of_mut, sync::atomic::compiler_fence};
 
+trait Word {}
+
+impl Word for u8 {}
+impl Word for u16 {}
+impl Word for u32 {}
+impl Word for i8 {}
+impl Word for i16 {}
+impl Word for i32 {}
+
+impl<W, const S: usize> crate::private::Sealed for [W; S] where W: Word {}
+
+impl<W, const S: usize> crate::private::Sealed for &[W; S] where W: Word {}
+
+impl<W> crate::private::Sealed for &[W] where W: Word {}
+
+impl<W> crate::private::Sealed for &mut [W] where W: Word {}
+
+/// Trait for buffers that can be given to DMA for reading.
+pub trait ReadBuffer: crate::private::Sealed {
+    /// Provide a buffer usable for DMA reads.
+    ///
+    /// The return value is:
+    ///
+    /// - pointer to the start of the buffer
+    /// - buffer size in bytes
+    ///
+    /// # Safety
+    ///
+    /// Once this method has been called, it is unsafe to call any `&mut self`
+    /// methods on this object as long as the returned value is in use (by DMA).
+    unsafe fn read_buffer(&self) -> (*const u8, usize);
```

Comment:
Since users won't every be dealing with the safety issues here (with the trait being sealed) is there much point to the `unsafe` keyword here?

Reply:
Keeping it as unsafe makes sense to me, especially since the DMA could be using the buffer at the point of calling it.